`type N {.importc: "const void *".} = pointer` creates order-dependent compilation failure

### Nim Version

```
Nim Compiler Version 2.2.10 [Linux: amd64]
Compiled at 2026-06-23
Copyright (c) 2006-2026 by Andreas Rumpf

git hash: bfeb3146d1638b39f69007a4ae5a23e23ae4e5ef
active boot switches: -d:release
```

```
Nim Compiler Version 2.2.11 [Linux: amd64]
Compiled at 2026-06-23
Copyright (c) 2006-2026 by Andreas Rumpf

git hash: 7b57dc1e54d2af08b12f995bb76cd44d663b3537
active boot switches: -d:release
```

```
Nim Compiler Version 2.3.1 [Linux: amd64]
Compiled at 2026-06-23
Copyright (c) 2006-2026 by Andreas Rumpf

git hash: f8e470eb579c96515693d76ad393dbd41b393f26
active boot switches: -d:release
```

### Description

```nim
type
  N {.importc: "const void *".} = pointer
  S = object
    f: proc (_: N) {.cdecl.}
#var _: proc (_: pointer) {.cdecl.}
proc d(_: proc (_: pointer) {.cdecl.}) = discard
discard S()
d(proc (_: pointer) {.cdecl.} = discard)
```

If one removes `discard S()`, it compiles. If one keeps `discard S()`, but then adds/uncomments `var _: proc (_: pointer) {.cdecl.}` , it also compiles.

The reason is that Nim treats `N` and `pointer` identically and folds them into the same typedef, which due to the `discard S()` gets forced into being the `const void *` form:
```c
typedef struct tyObject_S__jaD8Exa1LjcChMRWa9afMZw tyObject_S__jaD8Exa1LjcChMRWa9afMZw;
typedef N_CDECL_PTR(void, tyProc__32w2ZypAyMWuajvi2Jt79cg) (const void * __p0);
struct tyObject_S__jaD8Exa1LjcChMRWa9afMZw {
        tyProc__32w2ZypAyMWuajvi2Jt79cg f;
};
```

By adding back the `var _: proc (_: pointer) {.cdecl.}` or removing the `discard S()`, the same typedef reverts to being a non-const `void *`:
```c
typedef struct tyObject_S__jaD8Exa1LjcChMRWa9afMZw tyObject_S__jaD8Exa1LjcChMRWa9afMZw;
typedef N_CDECL_PTR(void, tyProc__32w2ZypAyMWuajvi2Jt79cg) (void* __p0);
struct tyObject_S__jaD8Exa1LjcChMRWa9afMZw {
        tyProc__32w2ZypAyMWuajvi2Jt79cg f;
};
```
at which point the `d(proc (_: pointer) {.cdecl.} = discard)` works.

None of these references to this `proc(pointer)` type have to particularly refer to each other, or even persist (`discard S()`, the `var _: proc (_: pointer) {.cdecl.}`); they just have to be the first one to trigger the creation of this `tyProc__32w2ZypAyMWuajvi2Jt79cg` `typedef`.

### Current Output

```text
CC: system/exceptions.nim
CC: std/private/digitsutils.nim
CC: system/dollars.nim
CC: system.nim
CC: k.nim
/tmp/tmp.DNht2FYTmb/@mk.nim.c: In function ‘NimMainModule’:
/tmp/tmp.DNht2FYTmb/@mk.nim.c:144:33: error: passing argument 1 of ‘d__k_u6’ from incompatible pointer type [-Wincompatible-pointer-types]
  144 |         nimln_(8);      d__k_u6(colonanonymous___k_u10);
      |                                 ^~~~~~~~~~~~~~~~~~~~~~
      |                                 |
      |                                 void (*)(void *)
/tmp/tmp.DNht2FYTmb/@mk.nim.c:80:72: note: expected ‘tyProc__32w2ZypAyMWuajvi2Jt79cg’ {aka ‘void (*)(const void *)’} but argument is of type ‘void (*)(void *)’
   80 | N_LIB_PRIVATE N_NIMCALL(void, d__k_u6)(tyProc__32w2ZypAyMWuajvi2Jt79cg __p0) {
      |                                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~
In file included from /tmp/tmp.DNht2FYTmb/@mk.nim.c:7:
/tmp/tmp.DNht2FYTmb/@mk.nim.c:84:29: note: ‘colonanonymous___k_u10’ declared here
   84 | N_LIB_PRIVATE N_CDECL(void, colonanonymous___k_u10)(void* __p0) {
```

### Expected Output

```text

```

### Known Workarounds

_No response_

### Additional Information

_No response_

--- a/compiler/sighashes.nim
+++ b/compiler/sighashes.nim
@@ -190,9 +190,10 @@ proc hashType(c: var MD5Context, t: PType; flags: set[ConsiderFlag]; conf: Confi
     if CoConsiderOwned in flags:
       c &= char(t.kind)
     c.hashType t.skipModifier, flags, conf
-  of tyBool, tyChar, tyInt..tyUInt64:
-    # no canonicalization for integral types, so that e.g. ``pid_t`` is
-    # produced instead of ``NI``:
+  of tyBool, tyChar, tyPointer, tyCstring, tyInt..tyUInt64:
+    # no canonicalization for builtin scalar-ish / pointer-like types, so
+    # that e.g. ``pid_t`` or an imported ``pointer`` alias keep their
+    # backend spelling instead of collapsing into the generic Nim builtin:
     c &= char(t.kind)
     if t.sym != nil and {sfImportc, sfExportc} * t.sym.flags != {}:
       c.hashSym(t.sym)
@@ -533,4 +534,3 @@ proc idOrSig*(s: PSym, currentModule: string,
     if counter != 0:
       result.add "_" & rope(counter+1)
     sigCollisions.inc(sig)
-

--- a/tests/ccgbugs/tsighash_typename_regression.nim
+++ b/tests/ccgbugs/tsighash_typename_regression.nim
@@ -30,3 +30,13 @@ block:
 
   var x = M(x: 1)
   doAssert(x.x == 1)
+
+block: # bug #25931
+  type
+    N {.importc: "const void *".} = pointer
+    S = object
+        f: proc (_: N) {.cdecl.}
+    #var _: proc (_: pointer) {.cdecl.}
+  proc d(_: proc (_: pointer) {.cdecl.}) = discard
+  discard S()
+  d(proc (_: pointer) {.cdecl.} = discard)

